File: solutions/4124-count-good-integers-on-a-grid-path/count-good-integers-on-a-grid-path.java

import java.util.*;

class Solution {
    long[][][] dp = new long[17][2][10];

    public long f(String s, List<Integer> indices, int i, boolean unequal, int prev2) {
        // Base case: If we successfully placed all 16 digits, we found 1 valid number
        if (i == 16) return 1;
        
        int unqState = unequal ? 1 : 0;
        if (dp[i][unqState][prev2] != -1) return dp[i][unqState][prev2];

        int start = 0;
        boolean flag = false;
        
        // Check if the current index in the 16-digit string lies on our target path
        for (int j = 0; j < indices.size(); j++) {
            if (indices.get(j) == i) {
                start = prev2; // Next digit MUST be >= previous path digit
                flag = true;
                break;
            }
        }

        long res = 0;
        // Maximum digit we can place based on the tight bound
        int mx = unequal ? 9 : (s.charAt(i) - '0');

        for (int j = start; j <= mx; j++) {
            int nextprev2 = flag ? j : prev2;
            
            // Recurse. The boundary becomes free if we pick a digit smaller than the max
            res += f(s, indices, i + 1, unequal || (j < mx), nextprev2);
        }

        return dp[i][unqState][prev2] = res;
    }

    public long get(long r, List<Integer> indices) {
        if (r < 0) return 0; // Edge case protection
        
        // Reset the DP table for the current limit calculation
        for (int i = 0; i < 17; i++) {
            for (int j = 0; j < 2; j++) {
                Arrays.fill(dp[i][j], -1);
            }
        }

        // Pad the number to exactly 16 digits
        String s = String.format("%016d", r);

        // Fast-forward through leading zeros
        int index = 0;
        for (int i = 0; i < 16; i++) {
            if (s.charAt(i) != '0') {
                index = i;
                break;
            }
        }

        int prev2 = 0;
        boolean flag = false;
        
        // Check if our starting non-zero index is a cell on the path
        for (int i = 0; i < indices.size(); i++) {
            if (indices.get(i) == index) {
                flag = true;
                break;
            }
        }

        int t = s.charAt(index) - '0';
        long ans = 0;

        // Manually process the first non-zero digit
        for (int i = 0; i <= t; i++) {
            int nextprev2 = flag ? i : prev2;
            
            if (i == t) {
                ans += f(s, indices, index + 1, false, nextprev2); // Prefix matches bound
            } else {
                ans += f(s, indices, index + 1, true, nextprev2); // Prefix is smaller
            }
        }
        return ans;
    }

    public long countGoodIntegersOnPath(long l, long r, String directions) {
        List<Integer> indices = new ArrayList<>();
        int row = 0, col = 0;
        
        indices.add(0); // Top-left cell
        
        // Convert 'D' and 'R' into 1D indices
        for (char c : directions.toCharArray()) {
            if (c == 'D') row++;
            else col++;
            indices.add(4 * row + col);
        }

        return get(r, indices) - get(l - 1, indices);
    }
}

File: solutions/4124-count-good-integers-on-a-grid-path/count-good-integers-on-a-grid-path.py
from functools import cache

class Solution:
    def countGoodIntegersOnPath(self, l: int, r: int, directions: str) -> int:
        indices = [0]
        row, col = 0, 0
        
        # Convert 'D' and 'R' directions into 1D grid indices
        for char in directions:
            if char == 'D':
                row += 1
            else:
                col += 1
            indices.append(4 * row + col)
            
        def get(limit: int) -> int:
            if limit < 0: 
                return 0
            
            # Pad the limit to exactly 16 digits
            s = str(limit).zfill(16)
            
            # Use Python's built-in memoization!
            @cache
            def f(i: int, unequal: bool, prev2: int) -> int:
                # Base case: successfully placed all 16 digits
                if i == 16: 
                    return 1
                
                # Check if current index lies on our target path
                flag = i in indices
                start = prev2 if flag else 0
                
                res = 0
                # Maximum digit we can place
                mx = 9 if unequal else int(s[i])
                
                for j in range(start, mx + 1):
                    nextprev2 = j if flag else prev2
                    # The boundary becomes unequal/free if we pick a digit smaller than max
                    res += f(i + 1, unequal or (j < mx), nextprev2)
                    
                return res

            # Fast-forward through leading zeros
            idx = 0
            for i in range(16):
                if s[i] != '0':
                    idx = i
                    break
                    
            prev2 = 0
            flag = idx in indices
            
            t = int(s[idx])
            ans = 0
            
            # Manually process the first non-zero digit to kick off the DP
            for i in range(t + 1):
                nextprev2 = i if flag else prev2
                
                if i == t:
                    ans += f(idx + 1, False, nextprev2) # Tight bound
                else:
                    ans += f(idx + 1, True, nextprev2)  # Free bound
            
            # Clear the cache to prevent interference between get(r) and get(l-1)
            f.cache_clear()
            return ans

        # Standard range DP calculation
        return get(r) - get(l - 1)